### DISCUSSION TITLE
send_file with > 500kb ends with MemoryError: memory allocation failed, allocating 1025 bytes

### DISCUSSION THREAD

**Discussion by Micha10** [miguelgrinberg/microdot#121] [Q&A] [answered]
Hi, 
I'm trying to send an angular FE-App to the browser. Some files are > 500kb. 

I've cross-compiled the latest microdot source into the micropython firmware and I'm using an esp8266 with 4MB

If I run my app and load the index.html the output is
```
Starting async server on 0.0.0.0:80...
GET / 200
static_route:  /runtime.js runtime.js
GET /runtime.js 200
static_route:  /main.js main.js
static_route:  /styles.css styles.css
static_route:  /polyfills.js polyfills.js
Task exception wasn't retrieved
future: <Task> coro= <generator object 'serve' at 3fff5020>
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "microdot_asyncio.py", line 269, in serve
  File "microdot_asyncio.py", line 337, in handle_request
  File "microdot_asyncio.py", line 155, in write
  File "microdot_asyncio.py", line 198, in __anext__
MemoryError: memory allocation failed, allocating 1025 bytes
Task exception wasn't retrieved
future: <Task> coro= <generator object 'serve' at 3fff68b0>
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "microdot_asyncio.py", line 269, in serve
  File "microdot_asyncio.py", line 337, in handle_request
  File "microdot_asyncio.py", line 155, in write
  File "microdot_asyncio.py", line 198, in __anext__
MemoryError: memory allocation failed, allocating 1025 bytes
GET /styles.css 200

```

Here are part's of the code
```
from microdot_asyncio import Microdot, send_file
web_app = Microdot()
...,
@web_app.route('/<path:path>')
        async def static(request, path):
            print("static_route: ", request.path, path)
            return send_file('/data/' + path)


uasyncio.run(self.__async_main())
async def __async_main(self):
        tasks = []
        task1 = uasyncio.create_task(self.do_other_things())
        task2 = uasyncio.create_task(self.webserver.init_webserver())
        tasks.append(task1)
        tasks.append(task2)
        await uasyncio.gather(*tasks)
        for task in tasks:
            await task
        print("Ende __async_main")
```

Is it a problem with the file size? As far as I read, micro. is sending big files in chunks.

Can you please give me a hint what's wrong?

It is possible that the browser is parallel-loading the needed files (main.js, styles.css, polyfills.js, runtime.js) and that there are 4 parallel requests to deliver files and that's too much for the esp?

**Comment by miguelgrinberg** [comment, marked as answer]
It cannot be a problem with file size, because as you've discovered, files are sent in fixed-size chunks. But if you get a few concurrent requests I guess it is possible that you are running out of memory.

Can you try adding this at the top of your server file?

```python
from microdot_asyncio import Response
Response.send_file_buffer_size = 256
```

The default is 1024 bytes for each file chunk. Hopefully you can find a size that allows all your static file requests to go through. Unfortunately the server can still blow up if two clients connect at the same time though.

**Reply by Micha10** [reply]
Thank you, that works without exception. Perhaps I'll try to lock the function so requests are answered synchron.

thanks for your fast help
